### PR TITLE
Add `get events` command to `irictl-machine`

### PR DESCRIPTION
**Pull request by ushabelgur**
# Proposed Changes

- Refactor event structure
- Refactor list event method in broker to fetch all events by kind `Machine` and API version
- Add get events command to irictl-machine

### CODE REVIEW

File: broker/machinebroker/server/event_list.go [hunk 1]
```diff
@@ -12,35 +12,56 @@ import (
...
+	InvolvedObjectAPIversionSelector = "involvedObject.apiVersion"
+)
 
-func (s *Server) listEvents(ctx context.Context, machineID string) ([]*iri.Event, error) {
+func (s *Server) listEvents(ctx context.Context) ([]*iri.Event, error) {
 	machineEventList := &v1.EventList{}
 	selectorField := fields.Set{}
-	selectorField[InvolvedObjectName] = machineID
+	selectorField[InvolvedObjectKindSelector] = InvolvedObjectKind
+	selectorField[InvolvedObjectAPIversionSelector] = computev1alpha1.SchemeGroupVersion.String()
```
Comment (lukasfrank) [inline broker/machinebroker/server/event_list.go:30]:
```suggestion
	selectorField := fields.Set{
		InvolvedObjectKindSelector:       InvolvedObjectKind,
		InvolvedObjectAPIversionSelector: computev1alpha1.SchemeGroupVersion.String(),
	}

	machineEventList := &v1.EventList{}
```

File: broker/machinebroker/server/event_list.go [hunk 2]
```diff
@@ -12,35 +12,56 @@ import (
...
+					Spec: &iri.EventSpec{
+						InvolvedObjectMeta: machineObjectMetadata,
+						Reason:             machineEvent.Reason,
+						Message:            machineEvent.Message,
+						Type:               machineEvent.Type,
+						EventTime:          eventTime,
+					},
+				}
+				iriEvents = append(iriEvents, iriEvent)
+			}
```
Comment (lukasfrank) [inline broker/machinebroker/server/event_list.go:63]: Can we flatten this by removing the else cases?

File: broker/machinebroker/server/event_list.go [hunk 3]
```diff
@@ -12,35 +12,56 @@ import (
...
-			Metadata: apiutils.GetIRIObjectMetadata(&machineEvent.ObjectMeta),
-			Spec: &iri.EventSpec{
-				Reason:    machineEvent.Reason,
-				Message:   machineEvent.Message,
-				Type:      machineEvent.Type,
-				EventTime: machineEvent.FirstTimestamp.Unix(),
-			},
+		ironcoreMachine, err := s.getIronCoreMachine(ctx, machineEvent.InvolvedObject.Name)
+		if err != nil {
+			continue
```
Comment (lukasfrank) [inline broker/machinebroker/server/event_list.go:43]: We might to log here an info that we were not able to fetch the machine

File: broker/machinebroker/server/event_list.go [hunk 4]
```diff
@@ -12,35 +12,56 @@ import (
...
+		} else {
+			machineObjectMetadata, err := apiutils.GetObjectMetadata(&ironcoreMachine.ObjectMeta)
+			if err != nil {
+				continue
+			} else {
+				eventTime = machineEvent.LastTimestamp.Unix()
+				if eventTime < 0 {
```
Comment (lukasfrank) [inline broker/machinebroker/server/event_list.go:50]: Is this check correct? Can you verify if the time is negative if it's not set?

**Comment by afritzler** [comment]
Can we rename the command to `list events` instead of using `get events`?

File: broker/machinebroker/server/event_list.go [hunk 5]
```diff
@@ -55,10 +76,14 @@ func (s *Server) filterEvents(events []*iri.Event, filter *iri.EventFilter) []*i
...
 			continue
 		}
-		if (filter.EventsFromTime >= 0 && iriEvent.Spec.EventTime >= filter.EventsFromTime) && (filter.EventsToTime >= 0 && iriEvent.Spec.EventTime <= filter.EventsToTime) {
+		if filter.EventsFromTime > 0 && filter.EventsToTime > 0 {
+			if iriEvent.Spec.EventTime >= filter.EventsFromTime && iriEvent.Spec.EventTime <= filter.EventsToTime {
+				res = append(res, iriEvent)
+			}
+		} else {
 			res = append(res, iriEvent)
 		}
```
Comment (lukasfrank) [inline broker/machinebroker/server/event_list.go:88]: Can we also streamline this that we only `append(res, iriEvent)`?

File: irictl-machine/cmd/irictl-machine/irictlmachine/get/get.go
Comment (lukasfrank) [inline irictl-machine/cmd/irictl-machine/irictlmachine/get/get.go:1]: Can we name this list events?

File: iri/apis/machine/v1alpha1/api.proto
Comment (lukasfrank) [inline iri/apis/machine/v1alpha1/api.proto:1]: Can we move the Event definition in a separate definition and reference it here?

**Review by lukasfrank — CHANGES_REQUESTED** [review]

File: broker/machinebroker/server/event_list.go (hunk 1 above)
Comment (ushabelgur) [reply broker/machinebroker/server/event_list.go]: done

File: broker/machinebroker/server/event_list.go (hunk 2 above)
Comment (ushabelgur) [reply broker/machinebroker/server/event_list.go]: removed unnecessary else blocks

File: broker/machinebroker/server/event_list.go (hunk 3 above)
Comment (ushabelgur) [reply broker/machinebroker/server/event_list.go:46]: done

File: broker/machinebroker/server/event_list.go (hunk 4 above)
Comment (ushabelgur) [reply broker/machinebroker/server/event_list.go]: I just rechecked k8s event recorder LastTimestamp is always set while generating event , so directly using it
https://github.com/kubernetes/kubernetes/blob/0fa4b9e14cba6342cfaaea4db793e36e4b8ae20d/staging/src/k8s.io/client-go/tools/record/event.go#L500

File: broker/machinebroker/server/event_list.go (hunk 5 above)
Comment (ushabelgur) [reply broker/machinebroker/server/event_list.go]: done

File: iri/apis/machine/v1alpha1/api.proto
Comment (ushabelgur) [reply iri/apis/machine/v1alpha1/api.proto:1]: Done

File: broker/machinebroker/server/event_list.go [hunk 6]
```diff
@@ -10,84 +10,91 @@ import (
...
 	"github.com/ironcore-dev/ironcore/broker/machinebroker/apiutils"
+	irievent "github.com/ironcore-dev/ironcore/iri/apis/event/v1alpha1"
 	iri "github.com/ironcore-dev/ironcore/iri/apis/machine/v1alpha1"
 )
 
-const InvolvedObjectName = "involvedObject.name"
+const (
+	InvolvedObjectKind               = "Machine"
+	InvolvedObjectKindSelector       = "involvedObject.kind"
+	InvolvedObjectAPIversionSelector = "involvedObject.apiVersion"
```
Comment (lukasfrank) [inline broker/machinebroker/server/event_list.go:25]:
```suggestion
	InvolvedObjectAPIVersionSelector = "involvedObject.apiVersion"
```

File: irictl-machine/cmd/irictl-machine/irictlmachine/get/get.go
Comment (lukasfrank) [reply irictl-machine/cmd/irictl-machine/irictlmachine/get/get.go:1]: Good catch! I would suggest to keep it like you implemented.

File: broker/machinebroker/server/event_list.go [hunk 7]
```diff
@@ -10,84 +10,91 @@ import (
...
-		res []*iri.Event
+		res []*irievent.Event
 		sel = labels.SelectorFromSet(filter.LabelSelector)
 	)
 	for _, iriEvent := range events {
-		if !sel.Matches(labels.Set(iriEvent.Metadata.Labels)) {
-			continue
-		}
-		if (filter.EventsFromTime >= 0 && iriEvent.Spec.EventTime >= filter.EventsFromTime) && (filter.EventsToTime >= 0 && iriEvent.Spec.EventTime <= filter.EventsToTime) {
+		if sel.Matches(labels.Set(iriEvent.Spec.InvolvedObjectMeta.Labels)) {
```
Comment (lukasfrank) [inline broker/machinebroker/server/event_list.go:76]: I think the time filter is only applied if the labels are matching however both should run after each other. How about? 
```
		if !sel.Matches(labels.Set(iriEvent.Spec.InvolvedObjectMeta.Labels)) {
			continue
		}

		if filter.EventsFromTime > 0 && filter.EventsToTime > 0 {
			if iriEvent.Spec.EventTime < filter.EventsFromTime || iriEvent.Spec.EventTime > filter.EventsToTime {
				continue
			}
		}

		res = append(res, iriEvent)

File: broker/machinebroker/server/event_list_test.go
Comment (lukasfrank) [inline broker/machinebroker/server/event_list_test.go:1]: It would be great to add tests to verify the correct filtering.

**Review by lukasfrank — COMMENTED** [review]
In `.reuse/dep5` `iri/apis/event` can be added to fix the failure of the pipeline: 

https://github.com/ironcore-dev/ironcore/actions/runs/9955097009/job/27502122368

File: broker/machinebroker/server/event_list_test.go
Comment (ushabelgur) [reply broker/machinebroker/server/event_list_test.go:1]: Added test cases to verify correct filtering

File: broker/machinebroker/server/event_list.go (hunk 7 above)
Comment (ushabelgur) [reply broker/machinebroker/server/event_list.go]: updated logic

File: broker/machinebroker/server/event_list.go (hunk 6 above)
Comment (ushabelgur) [reply broker/machinebroker/server/event_list.go]: done

**Review by lukasfrank — APPROVED** [review]
LGTM, great contribution!

**Pull request merged**